website/sales/views.py

```diff
@@ -1,12 +1,117 @@
 from django.contrib import messages
 from django.contrib.auth.decorators import login_required
+from django.core.exceptions import PermissionDenied
+from django.db import transaction
+from django.db.models import Q
 from django.shortcuts import get_object_or_404, redirect, render
+from django.urls import reverse
+from django.utils import timezone
 from django.utils.decorators import method_decorator
 from django.utils.translation import gettext_lazy as _
 from django.views import View
+from django.views.generic import DeleteView, DetailView
+from django.views.generic.edit import FormView
 
 from sales import services
-from sales.models.order import Order
+from sales.forms import ProductOrderForm
+from sales.models.order import Order, OrderItem, Shift
+
+
+@method_decorator(login_required, name="dispatch")
+class ShiftDetailView(DetailView):
+    model = Shift
+    context_object_name = "shift"
+    template_name = "sales/shift_view.html"
+
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+        shift = context["shift"]
+        queryset = Order.objects.filter(shift=shift).order_by("-created_at")
+        queryset = queryset.select_properties(
+            "total_amount", "subtotal", "num_items", "age_restricted"
+        )
+        queryset = queryset.prefetch_related(
+            "shift", "shift__event", "shift__product_list"
+        )
+        queryset = queryset.prefetch_related(
+            "order_items", "order_items__product", "order_items__product__product"
+        )
+        queryset = queryset.prefetch_related("payment")
+        context["orders"] = queryset.filter(
+            Q(payer=self.request.member) | Q(created_by=self.request.member)
+        )
+        return context
+
+
+@method_decorator(login_required, name="dispatch")
+class PlaceOrderView(FormView):
+    template_name = "sales/order_place.html"
+
+    def dispatch(self, request, *args, **kwargs):
+        self.shift = get_object_or_404(Shift, pk=kwargs["pk"])
+        if not self.shift.user_orders_allowed and request.method == "POST":
+            # Forbid POSTing when not in the correct time period
+            # or when only admins can manage this shift's orders
+            raise PermissionDenied
+        if not request.member.can_attend_events:
+            raise PermissionDenied
+        # TODO: if a shift belongs to an event, should we restrict self-ordering to event participants?
```

Comment:
Sounds wise. I think we don't currently do that for the food app. It's never used for registration-required events anyway, but it sounds sensible.